cmd/ndm_daemonset/controller/blockdevicestore.go

```diff
@@ -68,67 +71,49 @@ func (c *Controller) CreateBlockDevice(blockDevice apis.BlockDevice) error {
 		return err
 	}
 
-	/*
-	 * Update might failed due to to resource version mismatch which
-	 * can happen if some other entity updating same resource in parallel.
-	 */
-	err = c.UpdateBlockDevice(blockDevice, nil)
-	if err == nil {
-		return err
-	}
 	klog.Error("Update to blockdevice object failed: ", blockDevice.ObjectMeta.Name)
 	return nil
 }
 
+func boolPtr(b bool) *bool {
+	return &b
+}
+
 // UpdateBlockDevice update the BlockDevice resource in etcd
-func (c *Controller) UpdateBlockDevice(blockDevice apis.BlockDevice, oldBlockDevice *apis.BlockDevice) error {
+func (c *Controller) UpdateBlockDevice(blockDevice apis.BlockDevice) error {
 	var err error
 
-	blockDeviceCopy := blockDevice.DeepCopy()
-	if oldBlockDevice == nil {
-		oldBlockDevice = blockDevice.DeepCopy()
-		err = c.Clientset.Get(context.TODO(), client.ObjectKey{
-			Namespace: oldBlockDevice.Namespace,
-			Name:      oldBlockDevice.Name}, oldBlockDevice)
-		if err != nil {
-			klog.Errorf("eventcode=%s msg=%s : %v, err:%v rname=%v",
-				"ndm.blockdevice.update.failure",
-				"Failed to update block device : unable to get blockdevice object",
-				oldBlockDevice.ObjectMeta.Name, err, blockDeviceCopy.ObjectMeta.Name)
-			return err
-		}
-	}
-
-	blockDeviceCopy = mergeBlockDeviceData(*blockDeviceCopy, *oldBlockDevice)
+	klog.Infof("Blockdevice on API server %v", blockDevice)
+	patch := client.MergeFromWithOptions(blockDevice.DeepCopy())
 
-	err = c.Clientset.Update(context.TODO(), blockDeviceCopy)
+	err = c.Clientset.Patch(context.TODO(), &blockDevice, patch)
 	if err != nil {
 		klog.Errorf("eventcode=%s msg=%s : %v rname=%v",
 			"ndm.blockdevice.update.failure", "Unable to update blockdevice object",
-			err, blockDeviceCopy.ObjectMeta.Name)
+			err, blockDevice.ObjectMeta.Name)
 		return err
 	}
 	klog.Infof("eventcode=%s msg=%s rname=%v",
 		"ndm.blockdevice.update.success", "Updated blockdevice object",
-		blockDeviceCopy.ObjectMeta.Name)
+		blockDevice.ObjectMeta.Name)
 	return nil
 }
 
 // DeactivateBlockDevice API is used to set blockdevice status to "inactive" state in etcd
 func (c *Controller) DeactivateBlockDevice(blockDevice apis.BlockDevice) {
 
-	blockDeviceCopy := blockDevice.DeepCopy()
-	blockDeviceCopy.Status.State = NDMInactive
-	err := c.Clientset.Update(context.TODO(), blockDeviceCopy)
+	patch := client.MergeFrom(blockDevice.DeepCopy())
+	blockDevice.Status.State = NDMInactive
+	err := c.Clientset.Status().Patch(context.TODO(), &blockDevice, patch)
```

Comment:
Is this `client.Status().` required because we have status as a subresource?

Reply:
Yes

Reply:
Do we really need status as a subresource. Whats the best practice around that?

Reply:
These are the conventions: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status